docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,670 @@
+   while the rest is captured with the type variable `z`.
+
+Going back to **Example 5.2.1**, the set constructor would produce a set of
+equalities:
+
+```
+  a = Set(b)
+  a = Set(d)
+  b = Variant(RowCons("1a",
+                Rec(RowCons("tag", Str, RowCons("bal", Int, RowNil))),
+              c))
+  d = Variant(RowCons("2a",
+                Rec(RowCons("tag", Str,
+                  RowCons("bal", Int,
+                    RowCons("val", Int, RowNil)))),
+              e))
+```
+
+By solving these equalities with unification, we will arrive at the following
+variant:
+
+```
+  Set(Variant(RowCons(
+               "1a",
+               Rec(RowCons("tag", Str, RowCons("bal", Int, RowNil))),
+               RowCons(
+                 "2a",
+                 Rec(RowCons("tag", Str,
+                       RowCons("bal", Int,
+                         RowCons("val", Int, RowNil)))),
+                 t
+     ))))
+```
+
+Note that we still do not know the precise shape of the variant, as it
+closes with the type variable `t`. This is actually what we expect, as the set
+may be combined with records of other shapes. Normally, the final shape of a
+variant propagates via state variables of the TLA+ specification.
+
+**Type annotations for variants.** Snowcat requires that all state variables
+are annotated. What shall we write for variants?  We introduce the common type
+notation for variants that separates records with a pipe, that is, `|`. For
+instance, consider the following variable declaration in a TLA+ specification:
+
+```tla
+VARIABLES
+  \* @type: Set([ tag: "1a", bal: Int ] | [ tag: "2a", bal: Int, val: Int ]);
+  msgs
+```
+
+Note that even though the syntax of individual elements of a variant is very
+similar to that of a record, there is small difference: The tag field is not
+declared as a string type, but carries the values of the tag itself.
+
+As variants can grow large very quick, it is more convenient to introduce them
+via a type alias. For instance:
+
+```tla
+VARIABLES
+  (*
+    @typeAlias: MESSAGE =
+       [ tag: "1a", bal: Int ]
+     | [ tag: "2a", bal: Int, val: Int ]);
+    @type: Set(MESSAGE);
+   *)  
+  msgs
+```
+
+
+**Filter a set of variants.** As we have seen, the following pattern
+is quite common in TLA+ specifications, e.g., it is met in Paxos:
+
+```tla
+  \E m \in msgs:
+    /\ m.type = "1a"
+    ...
+  LET Q1b == { m \in msgs : m.type = "1b" /\ ... }
+```
+
+We introduce the operator `FilterByTag` that is a type-safe version of this
+pattern:
+
+```tla
+FilterByTag(Set, tag) == { e \in Set: e.tag = tag }
+```
+
+We introduce a special type inference rule for `FilterByTag`:
+
+```
+set: Set(Variant(RowCons("<TAG>", r, z)))
+------------------------------------------- [variant_filter]
+FilterByTag(set, "<TAG>"): Set(r)
+```
+
+Importantly, `FilterByTag` returns a set of records that carry the tag `<TAG>`,
+so we can access record fields of every individual record in the set.
```

Comment:
One important thing, which holds in practice (but not necessarily for an arbitrarily constructed type-term), but is not immediately obvious from this paragraph, is to emphasize that `r` is always of `Rec` shape, that is, it is always a complete record and not a generic row type. I'd make a special note on that point.

Reply:
This will make things more complicated on the SMT side. In this ADR, variants are just a type wrapper around records, so we do not any additional data structures to carry variants. If we allow arbitrary value, we would have to introduce another data structure for variants, which would be very similar to records.

Reply:
Right, I'd think to use a record to represent the variants, but allow variants themselves to hold any type. But this is not essential.